### PR TITLE
AP_NavEKF3: add GPS vAcc threshold that decides when to use the GPS as an altitude source

### PR DESCRIPTION
**Pull request by ntamas**
As discussed with @tridge last week in the EU dev call, this PR is the second in a series of changes that we'd like to upstream from our own fork of ArduPilot that we use for drone light shows to make it easier to maintain a parallel fork.

The purpose of this PR is to add an altitude source to the EKF3 (and also EKF2) that primarily uses GPS if the GPS is in RTK float or RTK fixed mode, falling back to barometric altitude if the GPS loses its RTK fix status.

I have to admit that the modifications turned out to be more extensive than what I initially anticipated - our fork simply modified the GPS altitude source in the EKF3 to fall back to baro if the GPS was not in RTK float / fix, while here I tried to be more structured and added a new EKF3 altitude source.

In retrospect, I would probably be happier if we could simply add a "minimum GPS fix status" parameter to the EKF3 parameters instead of adding a new EKF3 altitude source, so I'm putting up this PR for discussion in an upcoming dev call to get a common understanding of how I should proceed with this.

### CODE REVIEW

**Comment by rmackay9** [comment]
Hi @ntamas, txs very much for this.

I think one large question is where do we want to put the sensor source selection logic.  This essentially adds source selection logic into the EKF but so far at least we've always put that selection logic into lua scripts.  I always thought that we would eventually move some of that selection logic into C++ but I'm not sure it should be in the EKF.  I'm also not sure we should be blurring the lines between sensors enum and the sensor selection logic which is I think what's happening when we add a new sensor source enum value that is really a combination of two sensors.

The alternative approach would be to have two banks of sources (EK3_SRC1_xxx, EKF3_SRC2_xxx) and then have some C++ somewhere switching between the two if the GPS loses RTK lock.  That logic could be almost anywhere I think (vehicle c++ code, some new library, companion computer)

If we don't keep the selection logic separate from the enums, I worry that the logical next step would be to add a new GPS/VISO source and then add some params to specify the criteria to switch between them... and then a GPS/VISO/FLOW source, etc, etc.

**Comment by tridge** [comment]
@priseborough this would be a good one to discuss with you. I suspect a lot of fixed wing users would like this too. We just need to work out where the logic should sit

**Comment by rmackay9** [comment]
If we decide to add a parameter then we could add a EK3_GPS_USE_HGT to hold the threshold.. or something like that.

**Comment by WickedShell** [comment]
Out of curiosity is the target the fix type, or are we more interested in an accuracy threshold? For the most part I'm actually much more interested in using my GPS data when it passes a quality threshold rather then just fix based.

**Comment by ntamas** [comment]
I think the conclusion on today's dev call was to use some kind of a threshold based on the reported vertical accuracy of the GPS. I was advised to wait for a review from @priseborough and then most likely I'll rewrite the PR so it does not use a new enum member but a threshold parameter (unless there are objections).

**Comment by amilcarlucas** [comment]
I'm with @WickedShell on this

**Comment by ntamas** [comment]
Ready for another round of reviews. This time I've re-written the PR completely and added a new parameter (EK3_GPS_VACC_MAX) that acts as a threshold such that GPS is used for vertical position only if the reported vAcc value is below this threshold.

Some comments:

* Setting `EK3_GPS_VACC_MAX` to zero turns off the threshold completely, and this is the default to remain compatible with previous settings where the GPS was used as an altitude source unconditionally if it was designated as an altitude source in `EK3_SRC*_POSZ`.
* Hysteresis is added so we require 1s of a vAcc larger than the threshold to stop using the GPS as the altitude source and 3s of a vAcc smaller than the threshold to resume using the GPS. This is admittedly simpler than the lowpass + peak hold filter used for the sAcc, but I decided to use this because otherwise we would have needed to pick an arbitrary multiplier to the threshold by the user to decide when to resume using the GPS (the sAcc code uses 1.0 and 1.5 as the thresholds, but they are not user-configurable).
* On a somewhat unrelated note, it seems like the sAcc filter _accepts_ the estimated velocity if `dal.gps().speed_accuracy()` returns `false`, which is supposedly the case when the GPS provides no speed accuracy. Is that intentional, and if so, shall I do it the same way with the vertical accuracy? Right now I consider the vertical position accuracy test as failed if the GPS does not report a vAcc value.

**Comment by priseborough** [comment]
> * On a somewhat unrelated note, it seems like the sAcc filter _accepts_ the estimated velocity if `dal.gps().speed_accuracy()` returns `false`, which is supposedly the case when the GPS provides no speed accuracy. Is that intentional, and if so, shall I do it the same way with the vertical accuracy? Right now I consider the vertical position accuracy test as failed if the GPS does not report a vAcc value.

That was intentional and has worked OK for velocity because there is still the position and height accuracy metrics. For the use case introduced by this PR I would prefer to consider GPS height to be not useable if no accuracy is reported.

**Comment by priseborough** [comment]
> * Hysteresis is added so we require 1s of a vAcc larger than the threshold to stop using the GPS as the altitude source and 3s of a vAcc smaller than the threshold to resume using the GPS. This is admittedly simpler than the lowpass + peak hold filter used for the sAcc, but I decided to use this because otherwise we would have needed to pick an arbitrary multiplier to the threshold by the user to decide when to resume using the GPS (the sAcc code uses 1.0 and 1.5 as the thresholds, but they are not user-configurable).

The reason for the use of the peak hold and decay filter with sAcc was the noisy behaviour of that metric compared to hAcc and vAcc

**Review by priseborough — APPROVED** [review]
The hysteresis times may need to be adjusted following further use, but as is this reduces the likelihood of a loss of height control due to bad vertical GPS position.

**Comment by ntamas** [comment]
Rebased to latest `master` and resolved conflict in `AP_NavEKF3.h`.

**Pull request merged**
